File: app/src/main/java/com/futsch1/medtimer/medicine/MedicineViewAdapter.java
package com.futsch1.medtimer.medicine;


import android.os.Handler;
import android.os.HandlerThread;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.DiffUtil;

import com.futsch1.medtimer.database.FullMedicine;
import com.futsch1.medtimer.database.MedicineRepository;
import com.futsch1.medtimer.helpers.IdlingListAdapter;
import com.futsch1.medtimer.helpers.SwipeHelper;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MedicineViewAdapter extends IdlingListAdapter<FullMedicine, MedicineViewHolder> implements SwipeHelper.MovedCallback {

    private final HandlerThread thread;
    private final FragmentActivity activity;
    private final MedicineRepository medicineRepository;

    public MedicineViewAdapter(HandlerThread thread, FragmentActivity activity, @NotNull MedicineRepository medicineRepository) {
        super(new MedicineDiff());
        setHasStableIds(true);
        this.thread = thread;
        this.activity = activity;
        this.medicineRepository = medicineRepository;
    }

    // Create new views (invoked by the layout manager)
    @NonNull
    @Override
    public MedicineViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return MedicineViewHolder.create(parent, activity, thread);
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NonNull MedicineViewHolder holder, final int position) {
        FullMedicine current = getItem(position);
        holder.bind(current);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).medicine.medicineId;
    }

    @Override
    public void onMoved(int fromPosition, int toPosition) {
        List<FullMedicine> list = new ArrayList<>(getCurrentList());
        if (toPosition != -1) {
            Collections.swap(list, fromPosition, toPosition);
            submitList(list);
        }
    }

    @Override
    public void onMoveCompleted(int fromPosition, int toPosition) {
        if (fromPosition != toPosition) {
            new Handler(thread.getLooper()).post(() -> {
                        activity.runOnUiThread(() -> notifyItemMoved(fromPosition, toPosition));
                        medicineRepository.moveMedicine(fromPosition, toPosition);
                    }
            );
        }
    }

    public static class MedicineDiff extends DiffUtil.ItemCallback<FullMedicine> {

        @Override
        public boolean areItemsTheSame(@NonNull FullMedicine oldItem, @NonNull FullMedicine newItem) {
            return oldItem.medicine.medicineId == newItem.medicine.medicineId;
        }

        @Override
        public boolean areContentsTheSame(@NonNull FullMedicine oldItem, @NonNull FullMedicine newItem) {
            return oldItem.medicine.medicineId == newItem.medicine.medicineId &&
                    oldItem.medicine.name.equals(newItem.medicine.name) &&
                    oldItem.medicine.color == newItem.medicine.color &&
                    oldItem.medicine.iconId == newItem.medicine.iconId;
        }
    }

}

File: app/src/main/java/com/futsch1/medtimer/medicine/MedicineViewAdapter.kt
package com.futsch1.medtimer.medicine

import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DiffUtil
import com.futsch1.medtimer.database.FullMedicineEntity
import com.futsch1.medtimer.database.MedicineRepository
import com.futsch1.medtimer.di.Dispatcher
import com.futsch1.medtimer.di.MedTimerDispatchers
import com.futsch1.medtimer.helpers.IdlingListAdapter
import com.futsch1.medtimer.helpers.SwipeHelper.MovedCallback
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.launch
import java.util.Collections


class MedicineViewAdapter @AssistedInject constructor(
    @Assisted private val activity: FragmentActivity,
    private val medicineRepository: MedicineRepository,
    private val medicineViewHolderFactory: MedicineViewHolder.Factory,
    @param:Dispatcher(MedTimerDispatchers.IO) private val dispatcher: CoroutineDispatcher
) :
    IdlingListAdapter<FullMedicineEntity, MedicineViewHolder>(MedicineDiff()), MovedCallback {

    @AssistedFactory
    interface Factory {
        fun create(activity: FragmentActivity): MedicineViewAdapter
    }

    init {
        setHasStableIds(true)
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MedicineViewHolder {
        return medicineViewHolderFactory.create(parent, activity)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: MedicineViewHolder, position: Int) {
        val current = getItem(position)
        holder.bind(current)
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).medicine.medicineId.toLong()
    }

    override fun onMoved(fromPosition: Int, toPosition: Int) {
        if (toPosition != -1) {
            val list = currentList.toList()

            Collections.swap(list, fromPosition, toPosition)
            submitList(list)
        }
    }

    override fun onMoveCompleted(fromPosition: Int, toPosition: Int) {
        if (fromPosition != toPosition) {
            activity.lifecycleScope.launch(dispatcher) {
                medicineRepository.moveMedicine(fromPosition, toPosition)
            }
        }
    }

    class MedicineDiff : DiffUtil.ItemCallback<FullMedicineEntity>() {
        override fun areItemsTheSame(oldItem: FullMedicineEntity, newItem: FullMedicineEntity): Boolean {
            return oldItem.medicine.medicineId == newItem.medicine.medicineId
        }

        override fun areContentsTheSame(oldItem: FullMedicineEntity, newItem: FullMedicineEntity): Boolean {
            return oldItem == newItem
        }
    }
}